File: buildSrc/src/main/java/net/impactdev/gts/gradle/tasks/ReleaseNotesTask.java
/* This design originates from Nucleus, which is licensed under MIT */
package net.impactdev.gts.gradle.tasks;

import net.impactdev.gts.gradle.enums.ReleaseLevel;
import org.gradle.api.DefaultTask;
import org.gradle.api.GradleException;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.TaskAction;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;
import java.util.StringJoiner;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ReleaseNotesTask extends DefaultTask {

    @Internal
    public String result = "No templated release notes available";

    @Input public String version;
    @Input public String hash;
    @Input public String message;
    @Input public String sponge;
    @Input public ReleaseLevel level;

    public String getVersion() {
        return this.version;
    }

    public String getHash() {
        return this.hash;
    }

    private String getShortHash() {
        return this.hash.substring(0, 7);
    }

    public String getMessage() {
        return this.message;
    }

    public String getSponge() {
        return this.sponge;
    }

    public ReleaseLevel getLevel() {
        return this.level;
    }

    public String getResult() {
        return this.result;
    }

    @TaskAction
    public void run() {
        Path template = this.getProject().getRootDir().toPath()
                .resolve("changelogs")
                .resolve("templates")
                .resolve(this.getLevel().getTemplate() + ".md");


        if(Files.exists(template)) {
            try {
                String result = this.read(Files.newBufferedReader(template));
                this.result = Placeholders.transform(this, result);
            } catch (Exception e) {
                throw new GradleException(e.getMessage());
            }
        }

    }

    private Optional<String> getChangesForVersion() {
        Path notes = this.getProject().getRootDir().toPath()
                .resolve("changelogs")
                .resolve(this.getVersion() + ".md");

        if(Files.exists(notes)) {
            try {
                return Optional.of(this.read(Files.newBufferedReader(notes, StandardCharsets.UTF_8)));
            } catch (Exception e) {
                e.printStackTrace();
                return Optional.empty();
            }
        }

        return Optional.empty();
    }

    private String read(BufferedReader reader) throws IOException {
        StringBuilder result = new StringBuilder();
        String working;
        while((working = reader.readLine()) != null) {
            result.append(working).append('\n');
        }
        return result.toString();
    }

    private enum Placeholders {

        VERSION("version", ReleaseNotesTask::getVersion),
        HASH("commit-hash", ReleaseNotesTask::getHash),
        SHORT_HASH("commit-hash-short", ReleaseNotesTask::getShortHash),
        MESSAGE("commit-message", ReleaseNotesTask::getMessage),
        SPONGE("sponge", ReleaseNotesTask::getSponge),
        CHANGES("changes", task -> {
            Optional<String> result = task.getChangesForVersion();
            return result.orElse("No release notes available for this version...");
        }),
        ;

        private final String key;
        private final Function<ReleaseNotesTask, String> replacement;

        Placeholders(String key, Function<ReleaseNotesTask, String> replacement) {
            this.key = key;
            this.replacement = replacement;
        }

        public String getKey() {
            return this.key;
        }

        public Function<ReleaseNotesTask, String> getReplacement() {
            return this.replacement;
        }

        public static String transform(ReleaseNotesTask task, String in) {
            String out = in;
            for(Placeholders placeholder : Placeholders.values()) {
                out = out.replace("{{" + placeholder.getKey() + "}}", placeholder.getReplacement().apply(task));
            }

            return out;
        }
    }

}

File: buildSrc/src/main/kotlin/net/impactdev/gts/gradle/tasks/ReleaseNotesTask.kt
/* This design originates from Nucleus, which is licensed under MIT */
package net.impactdev.gts.gradle.tasks

import net.impactdev.gts.gradle.enums.ReleaseLevel
import org.gradle.api.DefaultTask
import org.gradle.api.GradleException
import org.gradle.api.tasks.TaskAction
import java.io.BufferedReader
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.util.*
import java.util.function.Function
import java.util.regex.Pattern

class ReleaseNotesTask : DefaultTask() {
    var result = "No templated release notes available"
    var version: String? = null
    var hash: String? = null
    var spongeVersion: String? = null
    var message: String? = null
    var sinceLastRelease: String? = null
    var level: ReleaseLevel? = null
    private val shortHash: String
        private get() = hash!!.substring(0, 7)

    @TaskAction
    fun run() {
        val template = project.rootDir.toPath()
            .resolve("changelogs")
            .resolve("templates")
            .resolve(level.getTemplate() + ".md")
        if (Files.exists(template)) {
            try {
                val result = read(Files.newBufferedReader(template))
                this.result = Placeholders.transform(this, result)
            } catch (e: Exception) {
                throw GradleException(e.message)
            }
        }
    }

    private val changesForVersion: Optional<String>
        private get() {
            val notes = project.rootDir.toPath()
                .resolve("changelogs")
                .resolve(version + ".md")
            return if (Files.exists(notes)) {
                try {
                    Optional.of(read(Files.newBufferedReader(notes, StandardCharsets.UTF_8)))
                } catch (e: Exception) {
                    e.printStackTrace()
                    Optional.empty<String>()
                }
            } else Optional.empty()
        }

    @Throws(IOException::class)
    private fun read(reader: BufferedReader): String {
        val result = StringBuilder()
        var working: String?
        while (reader.readLine().also { working = it } != null) {
            result.append(working).append('\n')
        }
        return result.toString()
    }

    private enum class Placeholders(val key: String, val replacement: Function<ReleaseNotesTask, String?>) {
        VERSION("version", Function { obj: ReleaseNotesTask -> obj.version }), HASH(
            "commit-hash",
            Function { obj: ReleaseNotesTask -> obj.hash }),
        SHORT_HASH(
            "commit-hash-short",
            Function<ReleaseNotesTask, String?> { obj: ReleaseNotesTask -> obj.shortHash }),
        MESSAGE("commit-message", Function { obj: ReleaseNotesTask -> obj.message }), SPONGE(
            "sponge",
            Function { obj: ReleaseNotesTask -> obj.spongeVersion }),
        CHANGES("changes", Function { task: ReleaseNotesTask ->
            val result = task.changesForVersion
            result.orElse("No release notes available for this version...")
        }),
        SINCE_LAST_RELEASE("since-last-release", Function<ReleaseNotesTask, String?> { task: ReleaseNotesTask ->
            val pattern = Pattern.compile("(?<commit>[0-9a-z]+)( [(].+[)])? (?<message>.+)")
            val matcher = pattern.matcher(task.sinceLastRelease)
            val joiner = StringJoiner("\n")
            while (matcher.find()) {
                var commit = matcher.group("commit")
                commit = "[" + commit.substring(0, 7) + "](https://github.com/NickImpact/GTS/commit/" + commit + ")"
                joiner.add(commit + " " + matcher.group("message") + "  ")
            }
            joiner.toString()
        });

        companion object {
            fun transform(task: ReleaseNotesTask, `in`: String): String {
                var out = `in`
                for (placeholder in values()) {
                    out = out.replace("{{" + placeholder.key + "}}", placeholder.replacement.apply(task))
                }
                return out
            }
        }
    }
}